README.md
# HTTP Client

A simple Net::HTTP wrapper with multi-part request, file upload and cookie-jar support.

## Install

```
gem install http-client
```

## Example

```ruby
require 'http-client'
res = HTTP::Client::Request.new(:get, "http://a.b/", max_redirects: 2).execute

# save a few keystrokes.
res = HTTP::Client.get("http://a.b/", max_redirects: 2)
res = HTTP::Client.put("http://a.b/", files: {pic: "kittens.jpg"}, query: {title: "the usual suspects"})
res = HTTP::Client.post("http://a.b/", auth: {username: "test", password: "test"}, headers: {'x-a' => 'b'})
res = HTTP::Client.delete("http://a.b/", auth: {username: "test", password: "test"});
```

## API

```
HTTP::Client
    .get(uri, arguments = {})
    .put(uri, arguments = {})
    .post(uri, arguments = {})
    .head(uri, arguments = {})
    .trace(uri, arguments = {})
    .delete(uri, arguments = {})
    .options(uri, arguments = {})

HTTP::Client
    .open_timeout=
    .read_timeout=
    .ssl_timeout=

HTTP::Client::Request
    .new(verb, uri, arguments = {})
    #execute

HTTP::Client::Response
    .new(net_http_response, last_effective_uri)
    #code
    #body
    #headers
    #last_effective_uri
```

### Exceptions

HTTP::Client rescues most Net::HTTP, Zlib and URI related exceptions and re-raises them as an appropriate exception.
You can retrieve the original exception using the `HTTP::Client::Error#original_error` instance method.

```
HTTP::Client::Error
  HTTP::Client::Error::URI
  HTTP::Client::Error::Zlib
  HTTP::Client::Error::Timeout
  HTTP::Client::Error::Transport
  HTTP::Client::Error::Argument
```

### Request parameters

#### Required

| Name | Type | Description |
|------|------|-------------|
| verb | Symbol | HTTP verb, one of :get, :head, :put, :post, :delete, :options, :trace. |
| uri | String or URI | Remote URI |

#### Optional arguments hash

| Name | Type | Description | Default |
|------|------|-------------|---------|
| headers | Hash | Net::HTTP headers, in key-value pairs. | nil |
| query | Hash | Net::HTTP query-string in key-value pairs. | nil |
| files | Hash | Multi-part file uploads, in key-value pairs of {name => path_to_file} or {name => File} | nil |
| body | String | Request body. | nil |
| auth | Hash | Basic-Auth hash. {username: "...", password: "..."} | nil |
| timeout | Integer | Fixed timeout for connection, read and ssl handshake in seconds. | Net::HTTP default |
| open_timeout | Integer | Connection timeout in seconds. | Net::HTTP default |
| read_timeout | Integer | Read timeout in seconds. | Net::HTTP default |
| ssl_timeout | Integer | SSL handshake timeout in seconds. | Net::HTTP default |
| max_redirects | Integer | Maximum redirects follow. | 0 |
| ssl_verify | Integer | OpenSSL verification. HTTP::Client::SSL_VERIFY_PEER or HTTP::Client::SSL_VERIFY_NONE | SSL_VERIFY_PEER |
| jar | HTTP::CookieJar | Optional cookie jar to use. Relies on HTTP::CookieJar from http-cookie gem. | HTTP::CookieJar.new |

## Default behaviour

### SSL

* By default peer verification is done. You can turn this off by passing `ssl_verfy: HTTP::Client::SSL_VERIFY_NONE` option.

### Redirects

* By default the client does not follow redirects. You can enable this with a non-zero `max_redirects` option.
* 301, 302 and 303 redirects are always followed with a GET method.
* 307 and 308 redirects preserve original request method & body.

## License

MIT


CHANGELOG
== 0.6.1 (2020-04-28)

* Bugfix - handles multi-part file uploads without other form data.

== 0.6.0 (2019-04-05)

* Accessors to set default global timeouts for Net::HTTP.

== 0.5.0 (2018-08-20)

* Handle basic auth credentials passed in URI.

== 0.4.0 (2017-07-05)

* Cleanup error classes namespacing.

== 0.3.0 (2017-06-30)

* Rescue error conditions and re-raise them as HTTP::Client::Error sub-classes.

== 0.2.2 (2016-09-21)

* URI parser error for relative paths.

== 0.2.1 (2016-09-21)

* Module scoped errors.

== 0.2.0 (2016-06-10)

* Uncompress gzip and deflate responses.

== 0.1.2 (2016-06-09)

* Fix YARD documentation.
* Preserve headers in 301, 302 & 303 redirects.
* Handle 303, 307 & 308.

== 0.1.1 (2016-06-08)

* Less restrictive dependency versions.
* Fix body assignment in requests.
* Tests

== 0.1.0 (2016-06-08)

* Initial version.


test/helper.rb
require 'bundler/setup'
require 'minitest/spec'
require 'minitest/autorun'
require 'minitest/reporters'
require 'server'
require 'http-client'

Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new



test/server.rb
require 'webrick'
require 'stringio'
require 'socket'
require 'uri'

class TestServer
  attr_reader :port

  def initialize app
    @app      = app
    @port     = find_port
  end

  def run path = '/'
    logfile   = StringIO.new
    logger    = WEBrick::Log.new logfile
    accesslog = [[logfile, WEBrick::AccessLog::COMBINED_LOG_FORMAT]]

    @server = WEBrick::HTTPServer.new Port: @port, Logger: logger, AccessLog: accesslog
    @server.mount_proc(path) do |req, res|
      code, headers, body = @app.call(req)
      res.status = code
      headers.each do |k, v|
        res[k] = v
      end
      res.body = [body].flatten.join
    end

    Thread.abort_on_exception = true
    @thread = Thread.new {@server.start}
  end

  def stop
    @thread && @thread.kill
  end

  def root
    URI.parse("http://127.0.0.1:#{port}/")
  end

  private
    def find_port
      @tcp = TCPServer.new('127.0.0.1', 0) rescue nil
      @tcp or raise RuntimeError 'Unable to find a local TCP port to bind.'
      port = @tcp.addr[1]
      @tcp.close
      port
    end
end


test/test_request.rb
require 'zlib'
require 'stringio'
require_relative 'helper'

describe 'HTTP Client Request' do
  it 'should reject invalid arguments' do
    assert_raises(HTTP::Client::Error::Argument, 'invalid verb') {HTTP::Client::Request.new(:foo, 'http://example.org/')}
    assert_raises(HTTP::Client::Error::URI, 'invalid uri') {HTTP::Client::Request.new(:get, 'http://')}
    assert_raises(HTTP::Client::Error::URI, 'invalid uri') {HTTP::Client::Request.new(:get, '/hello')}

    assert_raises(HTTP::Client::Error::Argument, 'invalid argument')  do
      HTTP::Client::Request.new(:get, 'http://example.org/', foo: 1)
    end
  end

  it 'validates body based on request verb' do
    assert_raises(HTTP::Client::Error::Argument, 'get cannot have body') do
      HTTP::Client::Request.new(:get, 'http://a.c', files: {test: __FILE__})
    end
  end

  it 'allows creation of valid request object' do
    assert HTTP::Client::Request.new(
      :post,
      'http://example.org/',
      query:         {title: 'test'},
      files:         {test1: __FILE__, test2: __FILE__},
      max_redirects: 2,
      timeout:       10,
      ssl_verify:    HTTP::Client::SSL_VERIFY_NONE,
      body:          'Hello world!'
    )
  end

  it 'executes a request and returns reponse' do
    app    = proc {|req| [200, {}, 'Hello World!']}
    server = TestServer.new(app)
    server.run

    response = HTTP::Client.get(server.root)
    server.stop
    assert response
    assert_kind_of Hash, response.headers
  end

  it 'raises timeout errors' do
    app    = proc {|req| sleep }
    server = TestServer.new(app)
    server.run
    assert_raises(HTTP::Client::Error::Timeout) {HTTP::Client.get(server.root, timeout: 0.2)}
  end

  it 'handles redirects' do
    cookie  = 'foo=bar; expires=Mon, 31-Dec-2038 23:59:59 GMT; path=/; domain=.127.0.0.1'

    [301, 302, 303, 307, 308]. each do |code|
      cookies = []
      methods = []
      headers = {}
      app     = proc do |req|
        cookies << req['cookie']
        case req.path
          when '/redirect2'
            req.each {|k,v| headers[k] = v}
            methods << req.request_method
            [200, {}, 'Hello World!']
          when '/redirect1'
            methods << req.request_method
            [code, {'Location' => '/redirect2'}, 'Moved!']
          when '/'
            [code, {'Location' => '/redirect1', 'Set-Cookie' => cookie}, 'Moved!']
        end
      end

      server = TestServer.new(app)
      server.run

      response = HTTP::Client.post(server.root)
      assert_equal code, response.code
      assert_equal %Q{#{server.root + '/redirect1'}}, response.headers['location']

      response = HTTP::Client.post(server.root, max_redirects: 1)
      assert_equal code, response.code
      assert_equal %Q{#{server.root + '/redirect1'}}, response.last_effective_uri.to_s
      assert_equal %Q{#{server.root + '/redirect2'}}, response.headers['location']

      response = HTTP::Client.post(server.root, headers: {'x-foo-1' => 'bar'}, max_redirects: 2)
      assert_equal 200, response.code
      assert_equal %Q{#{server.root + '/redirect2'}}, response.last_effective_uri.to_s

      server.stop
      cookies.compact!
      assert_equal 3, cookies.size
      assert_equal 1, cookies.uniq.size
      assert_equal 'foo=bar', cookies.first

      methods.compact!
      assert_equal 1, methods.uniq.size
      if [307, 308].include?(code)
        assert_equal 'POST', methods.first
      else
        assert_equal 'GET', methods.first
      end

      assert_equal 'bar',  headers['x-foo-1']
    end
  end

  it 'handles HTTP 1.1 redirects' do
    [307, 308].each do |code|
      bucket = []
      app    = proc do |req|
        case req.path
          when '/redirect'
            bucket = req.query
            [200, {}, 'Hello World!']
          when '/'
            [code, {'Location' => '/redirect'}, 'Moved!']
        end
      end

      server = TestServer.new(app)
      server.run

      response = HTTP::Client.post(server.root, query: {foo: 'bar'}, files: {baz: __FILE__}, max_redirects: 1)
      assert_equal 200, response.code
      assert_equal %Q{#{server.root + '/redirect'}}, response.last_effective_uri.to_s
      server.stop

      assert bucket['foo'], 'query 1 preserved'
      assert bucket['baz'], 'file upload preserved'
    end
  end

  it 'posts url encoded form' do
    qs = {}
    app = proc do |req|
      qs = req.query
      [200, {}, 'OK!']
    end
    server = TestServer.new(app)
    server.run

    response = HTTP::Client.post(server.root, query: {test1: 'test2'})
    server.stop
    assert_equal 200,     response.code
    assert_equal 'test2', qs['test1']
  end

  it 'posts multi-part encoded data' do
    qs = {}
    app = proc do |req|
      qs = req.query
      [200, {}, 'OK!']
    end
    server = TestServer.new(app)
    server.run

    response = HTTP::Client.post(server.root, files: {this: __FILE__}, query: {test1: 'test2'})
    server.stop

    assert_equal 200,     response.code
    assert_equal 'test2', qs['test1']

    data = IO.read(__FILE__)
    assert_equal data, qs['this']
    assert_equal File.basename(__FILE__), qs['this'].filename # people should stop overloading / extending String!
  end

  it 'handles gzip and deflate compression schemes.' do
    app = proc do |req|
      case req.path
        when '/gzip'
          [
            200,
            {'content-encoding' => 'gzip'},
            StringIO.new.tap {|io|
              gz = Zlib::GzipWriter.new(io)
              gz.write("Hello 1")
              gz.close
            }.string
          ]
        when '/deflate'
          [200, {'content-encoding' => 'deflate'}, Zlib.deflate("Hello 2")]
      end
    end

    server = TestServer.new(app)
    server.run

    response = HTTP::Client.get("#{server.root}/gzip", headers: {'accept-encoding' => 'gzip;deflate'})
    assert_equal 'Hello 1', response.body, 'gunzip ok'

    response = HTTP::Client.get("#{server.root}/deflate", headers: {'accept-encoding' => 'gzip;deflate'})
    assert_equal 'Hello 2', response.body, 'inflate ok'
    server.stop
  end
end


lib/http/client.rb
require 'securerandom'
require 'net/http'
require 'openssl'
require 'uri'
require 'mime/types'
require 'http-cookie'
require 'stringio'
require 'zlib'

module HTTP
  module Client
    VERSION = '0.6.1'

    GET                     = Net::HTTP::Get
    HEAD                    = Net::HTTP::Head
    PUT                     = Net::HTTP::Put
    POST                    = Net::HTTP::Post
    DELETE                  = Net::HTTP::Delete
    OPTIONS                 = Net::HTTP::Options
    TRACE                   = Net::HTTP::Trace

    SSL_VERIFY_NONE         = OpenSSL::SSL::VERIFY_NONE
    SSL_VERIFY_PEER         = OpenSSL::SSL::VERIFY_PEER

    class << self
      attr_accessor :open_timeout, :ssl_timeout, :read_timeout
    end

    class Request
      VALID_PARAMETERS        = %w(headers files query body auth timeout open_timeout ssl_timeout read_timeout max_redirects ssl_verify jar)
      DEFAULT_HEADERS         = {'User-Agent' => 'HTTP Client API/1.0'}

      REDIRECT_WITH_GET       = [301, 302, 303]
      REDIRECT_WITH_ORIGINAL  = [307, 308]

      VALID_VERBS             = [GET, HEAD, PUT, POST, DELETE, OPTIONS, TRACE]
      VALID_SSL_VERIFICATIONS = [SSL_VERIFY_NONE, SSL_VERIFY_PEER]
      VALID_REDIRECT_CODES    = REDIRECT_WITH_GET + REDIRECT_WITH_ORIGINAL

      # Create a new HTTP Client Request.
      #
      # @param  [Symbol]               verb            HTTP verb, one of :get, :head, :put, :post, :delete, :options, :trace.
      # @param  [String or URI]        uri             Remote URI.
      # @param  [Hash]                 args            Options, see below.
      # @option args [Hash]            headers         Net::HTTP headers, in key-value pairs.
      # @option args [Hash]            query           Net::HTTP query-string in key-value pairs.
      # @option args [Hash]            files           Multi-part file uploads, in key-value pairs of name & path or name & File object.
      # @option args [String]          body            Request body.
      # @option args [Hash]            auth            Basic-Auth hash. Requires :username and :password.
      # @option args [Integer]         timeout         Fixed timeout for connection, read and ssl handshake in seconds.
      # @option args [Integer]         open_timeout    Connection timeout in seconds.
      # @option args [Integer]         read_timeout    Read timeout in seconds.
      # @option args [Integer]         ssl_timeout     SSL handshake timeout in seconds.
      # @option args [Integer]         max_redirects   Max redirect follow, default: 0
      # @option args [Integer]         ssl_verify      OpenSSL verification, SSL_VERIFY_PEER (default) or SSL_VERIFY_NONE.
      # @option args [HTTP::CookieJar] jar             Optional cookie jar to use. Relies on HTTP::CookieJar from http-cookie gem.
      #
      # @return [HTTP::Client::Request]
      #
      # @example Retrieve a page using GET.
      #   request  = HTTP::Client::Request.new(:get, "http://www.example.org/", query: {q: "search something"})
      #   response = request.execute
      #
      # @example Handle redirects.
      #   request  = HTTP::Client::Request.new(:get, "http://www.example.org/", max_redirects: 3)
      #   response = request.execute
      #
      # @example Perform request and return result in one go.
      #   response = HTTP::Client.get("http://www.example.org/", max_redirects: 3)
      #
      # @example Upload a few files in a POST request.
      #   request  = HTTP::Client::Request.new(
      #     :post, "http://www.example.org/",
      #     files: {"cats" => "cats.jpg", "dogs" => "dogs.jpg"},
      #     query: {title: "cute pics"}
      #   )
      #   response = request.execute
      #
      # @example Pass in an external cookie jar.
      #   jar  = HTTP::CookieJar.new
      #   jar.load("mycookies.cky")
      #   response = HTTP::Client.get("http://www.example.org/", jar: jar)
      #
      def initialize verb, uri, args = {}
        args.each do |k, v|
          raise Error::Argument, "unknown argument #{k}" unless VALID_PARAMETERS.include?(k.to_s)
        end

        uri       = parse_uri!(uri)
        @delegate = create_request_delegate(verb, uri, args)

        if body = args[:body]
          raise Error::Argument, "#{verb} cannot have body" unless @delegate.class.const_get(:REQUEST_HAS_BODY)
          @delegate.body = body
        end

        # if auth is passed as arguments.
        if auth = args[:auth]
          @delegate.basic_auth(auth.fetch(:username), auth.fetch(:password))
        end

        # if auth is passed in uri.
        if uri.user && uri.password
          @delegate.basic_auth(uri.user, uri.password)
        end

        @open_timeout = HTTP::Client.open_timeout
        @read_timeout = HTTP::Client.read_timeout
        @ssl_timeout  = HTTP::Client.ssl_timeout

        # generic timeout
        if timeout = args[:timeout]
          @open_timeout = timeout
          @read_timeout = timeout
          @ssl_timeout  = timeout
        end

        # overrides
        @open_timeout = args[:open_timeout] if args[:open_timeout]
        @read_timeout = args[:read_timeout] if args[:read_timeout]
        @ssl_timeout  = args[:ssl_timeout]  if args[:ssl_timeout]

        @max_redirects = args.fetch(:max_redirects, 0)
        @ssl_verify    = args.fetch(:ssl_verify, SSL_VERIFY_PEER)
        @jar           = args.fetch(:jar, HTTP::CookieJar.new)
      end

      # Executes a request.
      #
      # @return [HTTP::Client::Response]
      #
      def execute
        last_effective_uri = uri

        cookie = HTTP::Cookie.cookie_value(@jar.cookies(uri))
        if cookie && !cookie.empty?
          @delegate.add_field('Cookie', cookie)
        end

        response = request!(uri, @delegate)
        @jar.parse(response['set-cookie'].to_s, uri)

        redirects = 0
        while redirects < @max_redirects && VALID_REDIRECT_CODES.include?(response.code.to_i)
          redirects         += 1
          last_effective_uri = parse_uri! response['location']
          redirect_delegate  = redirect_to(last_effective_uri, response.code.to_i)

          cookie = HTTP::Cookie.cookie_value(@jar.cookies(last_effective_uri))
          if cookie && !cookie.empty?
            redirect_delegate.add_field('Cookie', cookie)
          end

          response = request!(last_effective_uri, redirect_delegate)
          @jar.parse(response['set-cookie'].to_s, last_effective_uri)
        end

        Response.new(response, last_effective_uri)
      end

      private
        def uri
          @delegate.uri
        end

        def parse_uri! uri
          uri = uri.kind_of?(URI) ? uri : URI.parse(uri)
          case uri
            when URI::HTTP, URI::HTTPS
              raise Error::URI, "Invalid URI #{uri}" if uri.host.nil?
              uri
            when URI::Generic
              if @delegate && @delegate.uri
                @delegate.uri.dup.tap {|s| s += uri }
              else
                raise Error::URI, "Invalid URI #{uri}"
              end
            else
              raise Error::URI, "Invalid URI #{uri}"
          end
        rescue URI::InvalidURIError => e
          raise Error::URI, "Invalid URI #{uri}"
        end

        def create_request_delegate verb, uri, args
          klass   = find_delegate_class(verb)
          headers = DEFAULT_HEADERS.merge(args.fetch(:headers, {}))

          files    = args[:files]
          qs       = args[:query]
          uri      = uri.dup
          delegate = nil

          if files
            raise Error::Argument, "#{verb} cannot have body" unless klass.const_get(:REQUEST_HAS_BODY)
            multipart             = Multipart.new(files, qs)
            delegate              = klass.new(uri, headers)
            delegate.content_type = multipart.content_type
            delegate.body         = multipart.body
          elsif qs
            if klass.const_get(:REQUEST_HAS_BODY)
              delegate = klass.new(uri, headers)
              delegate.set_form_data(qs)
            else
              uri.query = URI.encode_www_form(qs)
              delegate  = klass.new(uri, headers)
            end
          else
            delegate = klass.new(uri, headers)
          end

          delegate
        end

        def request! uri, delegate
          http = Net::HTTP.new(uri.host, uri.port, :ENV)
          if uri.scheme == 'https'
            http.use_ssl     = true
            http.verify_mode = @ssl_verify
          end

          http.open_timeout = @open_timeout if @open_timeout
          http.read_timeout = @read_timeout if @read_timeout
          http.ssl_timeout  = @ssl_timeout  if @ssl_timeout

          response = http.request(delegate)
          http.finish if http.started?
          response
        rescue URI::Error => e
          raise Error::URI.new(e.message, e)
        rescue Zlib::Error => e
          raise Error::Zlib.new(e.message, e)
        rescue Timeout::Error => e
          raise Error::Timeout.new(e.message, e)
        rescue Net::HTTPBadResponse, Net::HTTPHeaderSyntaxError, Net::ProtocolError => e
          raise Error::Transport.new(e.message, e)
        end

        def redirect_to uri, code
          # NOTE request-uri with query string is not preserved.
          case code
            when *REDIRECT_WITH_GET
              GET.new(uri, {}).tap do |r|
                @delegate.each_header do |field, value|
                  next if field.downcase == 'host'
                  r[field] = value
                end
              end
            when *REDIRECT_WITH_ORIGINAL
              @delegate.class.new(uri, {}).tap do |r|
                @delegate.each_header do |field, value|
                  next if field.downcase == 'host'
                  r[field] = value
                end

                r.body = @delegate.body
              end
            else
              raise Error, "response #{code} should not result in redirection."
          end
        end

        def find_delegate_class verb
          if VALID_VERBS.include?(verb)
            verb
          else
            find_verb_class(verb.to_s)
          end
        end

        def find_verb_class string
          case string
            when /^get$/i     then GET
            when /^head$/i    then HEAD
            when /^put$/i     then PUT
            when /^post$/i    then POST
            when /^delete$/i  then DELETE
            else
              raise Error::Argument, "Invalid verb #{string}"
          end
        end
    end # Request

    class Response
      attr_reader :last_effective_uri, :response

      def initialize response, last_effective_uri
        @response           = response
        @last_effective_uri = last_effective_uri
      end

      def code
        response.code.to_i
      end

      def headers
        @headers ||= Hash[response.each_header.entries]
      end

      def body
        case headers['content-encoding'].to_s.downcase
          when 'gzip'
            gz = Zlib::GzipReader.new(StringIO.new(response.body))
            begin
              gz.read
            ensure
              gz.close
            end
          when 'deflate'
            Zlib.inflate(response.body)
          else
            response.body
        end
      end

      def inspect
        "#<#{self.class} @code=#{code} @last_effective_uri=#{last_effective_uri}>"
      end
    end # Response

    class Multipart
      attr_reader :boundary

      EOL               = "\r\n"
      DEFAULT_MIME_TYPE = 'application/octet-stream'

      def initialize files, query = {}
        @files    = files
        @query    = query || {}
        @boundary = generate_boundary
      end

      def content_type
        "multipart/form-data; boundary=#{boundary}"
      end

      def body
        body      = ''.encode('ASCII-8BIT')
        separator = "--#{boundary}"

        if @query && !@query.empty?
          @query.each do |key, value|
            body << separator << EOL
            body << %Q{Content-Disposition: form-data; name="#{key}"} << EOL
            body << EOL
            body << value
            body << EOL
          end
        end

        if @files && !@files.empty?
          @files.each do |name, handle|
            if handle.respond_to?(:read)
              path = handle.path
              data = io.read
            else
              path = handle
              data = IO.read(path)
            end

            filename = File.basename(path)
            mime     = mime_type(filename)

            body << separator << EOL
            body << %Q{Content-Disposition: form-data; name="#{name}"; filename="#{filename}"} << EOL
            body << %Q{Content-Type: #{mime}}              << EOL
            body << %Q{Content-Transfer-Encoding: binary}  << EOL
            body << %Q{Content-Length: #{data.bytesize}}   << EOL
            body << EOL
            body << data
            body << EOL
          end
        end

        body << separator << "--" << EOL
        body
      end

      private
        def generate_boundary
          SecureRandom.random_bytes(16).unpack('H*').first
        end

        def mime_type filename
          MIME::Types.type_for(File.extname(filename)).first || DEFAULT_MIME_TYPE
        end
    end # Multipart

    # Helpers
    class << self
      # Creates a GET request and executes it, returning the response.
      # @see HTTP::Client::Request#initialize
      #
      # @param  [String or URI] uri   Remote URI.
      # @param  [Hash]          args  Options, see HTTP::Client::Request#initialize.
      # @return [HTTP::Client::Response]
      #
      def get uri, args = {}; Request.new(GET, uri, args).execute; end

      # Creates a PUT request and executes it, returning the response.
      # @see HTTP::Client::Request#initialize
      #
      # @param  [String or URI] uri   Remote URI.
      # @param  [Hash]          args  Options, see HTTP::Client::Request#initialize.
      # @return [HTTP::Client::Response]
      #
      def put uri, args = {}; Request.new(PUT, uri, args).execute; end

      # Creates a POST request and executes it, returning the response.
      # @see HTTP::Client::Request#initialize
      #
      # @param  [String or URI] uri   Remote URI.
      # @param  [Hash]          args  Options, see HTTP::Client::Request#initialize.
      # @return [HTTP::Client::Response]
      #
      def post  uri, args = {}; Request.new(POST, uri, args).execute; end

      # Creates a DELETE request and executes it, returning the response.
      # @see HTTP::Client::Request#initialize
      #
      # @param  [String or URI] uri   Remote URI.
      # @param  [Hash]          args  Options, see HTTP::Client::Request#initialize.
      # @return [HTTP::Client::Response]
      #
      def delete uri, args = {}; Request.new(DELETE, uri, args).execute; end

      # Creates a OPTIONS request and executes it, returning the response.
      # @see HTTP::Client::Request#initialize
      #
      # @param  [String or URI] uri   Remote URI.
      # @param  [Hash]          args  Options, see HTTP::Client::Request#initialize.
      # @return [HTTP::Client::Response]
      #
      def options uri, args = {}; Request.new(OPTIONS, uri, args).execute; end

      # Creates a TRACE request and executes it, returning the response.
      # @see HTTP::Client::Request#initialize
      #
      # @param  [String or URI] uri   Remote URI.
      # @param  [Hash]          args  Options, see HTTP::Client::Request#initialize.
      # @return [HTTP::Client::Response]
      #
      def trace uri, args = {}; Request.new(TRACE, uri, args).execute; end
    end

    class Error < StandardError
      attr_reader :original_error

      def initialize message, original_error = nil
        @original_error = original_error
        super(message)
      end

      class URI < Error; end
      class Zlib < Error; end
      class Timeout < Error; end
      class Transport < Error; end
      class Argument < Error; end
    end # Error
  end # Client
end # HTTP
